src/main_macro.rs

```diff
@@ -29,23 +31,68 @@
 ///     println!("{}", x);
 /// });
 /// ```
+///
+/// With command line arguments.
+///
+/// ```rust,ignore
+/// #[macro_use] extern crate quicli;
+/// use quicli::prelude::*;
+/// use std::path::PathBuf;
+///
+/// #[derive(Debug, StructOpt)]
+/// struct Cli {
+///     /// The config file for the cli app.
+///     #[structopt(long = "config", short = "c", parse(from_os_str))]
+///     config: PathBuf,
+/// }
+///
+/// main!(|args: Cli| {
+///     let x = read_file(&args.config)?;
+///     println!("{}", x);
+/// });
+/// ```
+///
+/// With command line arguments including a verbosity flag.
+///
+/// ```rust,ignore
+/// #[macro_use] extern crate quicli;
+/// use quicli::prelude::*;
+/// use std::path::PathBuf;
+///
+/// #[derive(Debug, StructOpt)]
+/// struct Cli {
+///     /// The config file for the cli app.
+///     #[structopt(long = "config", short = "c", parse(from_os_str))]
+///     config: PathBuf,
+///     #[structopt(flatten)]
+///     verbosity: Verbosity,
+/// }
+///
+/// main!(|args: Cli, log_level: Verbosity| {
+///     let x = read_file(&args.config)?;
+///     println!("{}", x);
+/// });
+/// ```
 #[macro_export]
 macro_rules! main {
     (|$args:ident: $cli:ty, log_level: $verbosity:ident| $body:expr) => {
         fn main() {
-            fn run() -> $crate::prelude::Result<()> {
-                let $args = <$cli>::from_args();
-                $args.$verbosity.setup_env_logger(&env!("CARGO_PKG_NAME"))?;
-
+            #[allow(unused_variables)]
+            fn run($args: $cli) -> $crate::prelude::Result<()> {
                 $body
-
+                #[allow(unreachable_code)]
                 Ok(())
             }
 
-            if let Err(e) = run() {
-                eprintln!("error: {}", e);
+            let $args = <$cli>::from_args();
+            // This cannot fail because we control the main method, so no-one else can register a
+            // log handler before us.
+            $args.$verbosity.setup_env_logger(&env!("CARGO_PKG_NAME"))
+                .expect("logger setup should never fail, this is a bug in quicli");
```

Comment:
Can you expand this expect a bit to be something like 

```rust
if let Err(e) = $args.$verbosity.setup_env_logger(&env!("CARGO_PKG_NAME")) {
    eprintln!("Failed to set up log handler. This is a bug in quicli!");
    eprintln!("Please report it: https://github.com/killercup/quicli/issues");
    eprintln!("Details: {}", e);
    ::std::process::exit(1);
}
```

(Same below)

Reply:
I dont' mind doing this, but it really should never fail, as the only way a logger init can fail is if there already is a logger installed, and we know that there isn't in this case.

I just included the `expect` as a minimal sanity check - I should probably just use `unwrap`.